Update docs (#465)
megaease/easegress: doc/developer-guide.md

- [Developer Guide](#developer-guide)
	- [Architecture](#architecture)
	- [Project Layout](#project-layout)
	- [Building and testing](#building-and-testing)
- [Extending Easegress](#extending-easegress)
	- [Developing an Object](#developing-an-object)
		- [Main Business Logic](#main-business-logic)
		- [Register Object to Supervisor](#register-object-to-supervisor)
	- [Developing a Filter](#developing-a-filter)
		- [Main Business Logic](#main-business-logic-1)
		- [Register Filter to Pipeline](#register-filter-to-pipeline)
		- [JumpIf Mechanism in Pipeline](#jumpif-mechanism-in-pipeline)

## Architecture

The following diagram describes the high-level architecture of Easegress. The Cluster component is responsible for synchronizing data between of different nodes or after restarts. The Supervisor manages the lifecycle of all kinds of objects.

![architecture](./imgs/architecture.png)

Easegress has four different *kind* of objects:

> 1. **System Controllers** operate essential system level activities. Every instance of Easegress has exactly one instance of each kind of System Controllers.

Reviewer: ```suggestion
1. **System Controllers** operate essential system-level activities. Every instance of Easegress launches exactly one instance of each kind of System Controllers after the very start.
```

> 2. **Business Controllers** can be created and deleted using admin operations. Business Controller operate mainly other tasks than handling directly the traffic.

Reviewer: ```suggestion
2. **Business Controllers** can be created and deleted using admin operations. Business Controllers operate mainly on other tasks than handling directly the traffic.
```

```

### Register Filter to Pipeline

Our core logic is very simple, now let's add some non-business code to make our new filter conform with the requirement of the Pipeline framework. All filters must satisfy the interface `Filter` in [`pkg/object/httppipeline/registry.go`](https://github.com/megaease/easegress/blob/master/pkg/object/httppipeline/registry.go).

> All of the methods with their names and comments are clean, the only one we need to emphasize is `Inherit`. It is called when the pipeline is updated, without modifying the filters identity (*name* and *kind*). In practice this happens when underlying machine restarts and Easegress. It's the filter's own responsibility to do hot-update in `Inherit` such as transferring meaningful consecutive data.

Reviewer: ```suggestion
All of the methods with their names and comments are clean, the only one we need to emphasize is `Inherit`. It is called when the pipeline is updated, without modifying the filter's identity (*name* and *kind*). In practice, this happens when the underlying machine restarts and Easegress. It's the filter's own responsibility to do hot-update in `Inherit` such as transferring meaningful consecutive data.
```

Author: https://github.com/megaease/easegress/pull/465/commits/c845feec4004b9f304cd5b32c641b8a654793d36
